shiny/render/_dataframe.py

```diff
@@ -258,29 +402,334 @@ class data_frame(Renderer[DataFrameResult]):
+                    # prev = update_info["prev"]
+
+                    formatted_value = await self._cell_update_fn(info=update_info)
+                    #     row_index=row_index,
+                    #     column_index=column_index,
+                    #     value=value,
+                    #     prev=prev,
+                    # )
+                    # TODO-barret; check type here?
+                    # TODO-barret; The return value should be coerced by pandas to the correct type
+                    formatted_values.append(formatted_value)
+
+                return formatted_values
+
+        self.set_cell_update_fn(_set_cell_update_default)
+        self.set_cells_update_fn(_set_cells_update_default)
+        # self._add_message_handlers()
+
+    # To be called by session's output_handler message handler on this data_frame instance
+    @output_dispatch_handler
+    # Do not change this method name unless you update corresponding code in `/js/dataframe/`!!
+    async def _handle_cells_update(self, update_infos: list[CellUpdateInfo]):
+
+        # Make new array to trigger reactive update
+        patches = [p for p in self.cell_patches()]
+
+        # Call user's cell update method to retrieve formatted values
+        updated_raw_values = await self._cells_update_fn(infos=update_infos)
+
+        if (not isinstance(updated_raw_values, list)) or len(updated_raw_values) != len(
+            update_infos
+        ):
+            raise ValueError(
+                f"The return value of {self.output_id}'s `handle_cells_update()` "
+                f"(typically set by `@{self.output_id}.set_cells_update_fn`) "
+                "must be a list of the same length as the input list of cell updates. "
+                f"Received {len(updated_raw_values)} items and expected {len(update_infos)}."
+            )
+
+        # Add new patches
+        for updated_raw_value, update_info in zip(updated_raw_values, update_infos):
+            patches.append(
+                CellPatch(
+                    row_index=update_info["row_index"],
+                    column_index=update_info["column_index"],
+                    value=updated_raw_value,
+                    prev=update_info["prev"],
+                )
+            )
+
+        # Remove duplicate patches
+        patch_map: dict[tuple[int, int], CellPatch] = {}
+        for patch in patches:
+            patch_map[(patch.row_index, patch.column_index)] = patch
+        patches = list(patch_map.values())
+
+        # Set new patches
+        self.cell_patches.set(patches)
+
+        return updated_raw_values
 
     def auto_output_ui(self) -> Tag:
         return ui.output_data_frame(id=self.output_id)
 
-    async def transform(self, value: DataFrameResult) -> Jsonifiable:
+    def __init__(self, fn: ValueFn[DataFrameResult]):
+        # Must be done before super().__init__ is called
+        session = get_current_session()
+        self._session = session
```

Comment:
Expand on the comment: why does this need to be done before `super().__init__()`?